networkx/classes/digraph.py

```diff
@@ -457,28 +463,15 @@ def add_nodes_from(self, nodes_for_adding, **attr):
 
         """
         for n in nodes_for_adding:
-            # keep all this inside try/except because
-            # CPython throws TypeError on n not in self._succ,
-            # while pre-2.7.5 ironpython throws on self._succ[n]
             try:
-                if n not in self._succ:
-                    self._succ[n] = self.adjlist_inner_dict_factory()
-                    self._pred[n] = self.adjlist_inner_dict_factory()
-                    self._node[n] = attr.copy()
-                else:
-                    self._node[n].update(attr)
+                hash(n)
```

Comment:
You are only testing for whether the node is hashable, but this code is supposed to check if the node is already in the DiGraph.

Reply:
old code 
```
try:
    if n not in self._succ:
        self._succ[n]
```
this code tried to add|update node.

but if exception raised 
```
nn, ndict = n
```
this code try to unpack (node, data) tuple and after  
```
if nn not in self._succ:
```
this code is duplicate of previous try block.

hash(n) is more simple check IMO.